wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginChainBuilder.java

```diff
@@ -0,0 +1,229 @@
+          throw new SQLException(
+              Messages.get(
+                  "ConnectionPluginManager.unknownPluginCode",
+                  new Object[] {pluginCode}));
+        }
+        pluginFactories.add(pluginFactoriesByCode.get(pluginCode));
+      }
+    }
+
+    if (!pluginFactories.isEmpty()) {
+
+      if (!PropertyDefinition.PRESERVE_PLUGINS_ORDER.getBoolean(props)) {
+        pluginFactories = this.sortPluginFactories(pluginFactories);
+
+        final List<Class<? extends ConnectionPluginFactory>> tempPluginFactories = pluginFactories;
+        LOGGER.finest(() ->
+            "Plugins order has been rearranged. The following order is in effect: "
+                + tempPluginFactories.stream()
+                  .map(Class::getSimpleName)
+                  .collect(Collectors.joining(", ")));
+      }
+
+      try {
+        final ConnectionPluginFactory[] factories =
+            WrapperUtils.loadClasses(
+                    pluginFactories,
+                    ConnectionPluginFactory.class,
+                    "ConnectionPluginManager.unableToLoadPlugin")
+                .toArray(new ConnectionPluginFactory[0]);
+
+        // make a chain of connection plugins
+
+        plugins = new ArrayList<>(factories.length + 1);
+
+        for (final ConnectionPluginFactory factory : factories) {
+          plugins.add(factory.getInstance(pluginService, props));
+        }
+
+      } catch (final InstantiationException instEx) {
+        throw new SQLException(instEx.getMessage(), SqlState.UNKNOWN_STATE.getState(), instEx);
+      }
+    } else {
+      plugins = new ArrayList<>(1); // one spot for default connection plugin
+    }
+
+    // add default connection plugin to the tail
+    final ConnectionPlugin defaultPlugin =
+        new DefaultConnectionPlugin(pluginService, defaultConnProvider, pluginManagerService);
+    plugins.add(defaultPlugin);
+
+    return plugins;
+  }
+
+  protected List<Class<? extends ConnectionPluginFactory>> sortPluginFactories(
+      final List<Class<? extends ConnectionPluginFactory>> unsortedPluginFactories) {
+
+    ArrayList<PluginFactoryInfo> weights = new ArrayList<>();
+    for (Class<? extends ConnectionPluginFactory> pluginFactory : unsortedPluginFactories) {
+      Integer pluginFactoryWeight = pluginWeightByPluginFactory.get(pluginFactory);
+      if (pluginFactoryWeight == null) {
+        // This plugin factory is unknown. It might be a user custom plugin factory.
+        // Let's keep this plugin factory original position.
+        pluginFactoryWeight = STICK_TO_PRIOR_PLUGIN;
+      }
+      weights.add(new PluginFactoryInfo(pluginFactory, pluginFactoryWeight));
+    }
+
+    // Determine weights for plugin factories with STICK_TO_PRIOR_PLUGIN
+    int lastWeight = 0;
+    for (PluginFactoryInfo info : weights) {
+      if (info.weight != STICK_TO_PRIOR_PLUGIN) {
+        lastWeight = info.weight;
+        continue;
+      }
+      lastWeight++;
```

Comment:
this bit is a bit confusing - why do we need to increment and reassign weights?

Reply:
Reworked this piece. Replaced two loops with a single one. Added more comments to explain the algorithm.